### PR TITLE
Adding VectorCAST coverage

### PR DESCRIPTION
**Pull request by TimSVector**
Adding MC/DC, Function, and Function Call coverage.  Results being shown in the coverage report and associated graphs.  Method coverage not working for some reason.

### Testing done

Tested with an expanded Cobertura xml file.  Run through multiple configurations on Jenkins server with varying coverage configurations

```[tasklist]
### Submitter checklist
- [X] Open a fork of the repository
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Comment by TimSVector** [comment]
Not sure how to fix this the missing comma affer "id" : "error-prone"

```
Quality Monitor PR / Build, test and monitor quality on Ubuntu (pull_request_target) Failing after 3m

"tools": [
  {
    "id": "spotbugs",
    "sourcePath": "src/main/java",
    "pattern": "**/target/spotbugsXml.xml"
  },
  {
    "id": "error-prone"
    "pattern": "**/maven.log"
  }
```

File: src/main/java/edu/hm/hafner/coverage/Metric.java [hunk 1]
```diff
@@ -111,7 +116,10 @@ public static NavigableSet<Metric> getCoverageMetrics() {
                 METHOD,
                 LINE,
                 BRANCH,
-                INSTRUCTION
+                INSTRUCTION,
+                MCDC_PAIR,
+                FUNCTION,
+                FUNCTION_CALL
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/coverage/Metric.java:122]: I haven't found time to look into your sources yet (3 other parsers are in the queue 😄 ).

What I do not yet understand: what is the meaning of each of these coverage metrics? And how do these metrics relate to the existing ones?

Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:122]: For the different coverages:
- Function Coverage: This should be the same as Method Coverage (ie. did this function get called at all).  I was was unable to get the function coverage metric to translate into the existing Method Coverage metric.
- Function Call Coverage: This is a measure of the number of function calls in a function/file that have been made.  So if one has 3 function calls in a particular function, have those function calls been executed
- Modified Condition/Decision Coverage (MCDC) is a type of code coverage that includes equivalence class and boundary value coverage. **MCDC pairs are used to test for independence, which means that each condition must independently affect the decision outcome**. For example, in the truth table approach, test cases that provide MC/DC are selected by identifying pairs of rows where only one condition and the decision outcome change values between the two rows. 

**Function & Function Call** coverages are required for ISO-26262 Architectural Level testing 
**MCDC Equivalence Pair** coverage is required for for DO-178B/C Level A, the highest level of ISO 26262, and other standards.

Here's a breakdown of where VectorCAST generates coverages per the standards our customers use

Screenshot 2024-05-14 134337

**Comment by TimSVector** [comment]
> Not sure how to fix this the missing comma affer "id" : "error-prone"
> 
> ```
> Quality Monitor PR / Build, test and monitor quality on Ubuntu (pull_request_target) Failing after 3m
> 
> "tools": [
>   {
>     "id": "spotbugs",
>     "sourcePath": "src/main/java",
>     "pattern": "**/target/spotbugsXml.xml"
>   },
>   {
>     "id": "error-prone"
>     "pattern": "**/maven.log"
>   }
> ```

@uhafner - do you know how to clear this error?  I don't know what I've done in the branch to create it.  I'd like to move on to the task of fixing the coverage-plugin changes that are pending this artifact

**Comment by uhafner** [comment]
I think I fixed that in main.

**Comment by uhafner** [comment]
It seems that you did commit your changes in main. This is not ideal. You should always create changes in a separate branch. Then it is much easier to update from the upstream repository.

**Comment by TimSVector** [comment]
I thought you requested a fork and a PR.  I can redo if you’d like.
________________________________

File: src/main/java/edu/hm/hafner/coverage/Metric.java (hunk 1 above)
Comment (uhafner) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:122]: > For the different coverages:
> 
> * Function Coverage: This should be the same as Method Coverage (ie. did this function get called at all).  I was was unable to get the function coverage metric to translate into the existing Method Coverage metric.

I see. I will have a look why this is not possible...

> * Function Call Coverage: This is a measure of the number of function calls in a function/file that have been made.  So if one has 3 function calls in a particular function, have those function calls been executed

Ok, do you have an example in your tests? I did not find such a report? All reports use `functioncall-coverage` only on the class nodes but not in the line nodes.

> * Modified Condition/Decision Coverage (MCDC) is a type of code coverage that includes equivalence class and boundary value coverage. **MCDC pairs are used to test for independence, which means that each condition must independently affect the decision outcome**. For example, in the truth table approach, test cases that provide MC/DC are selected by identifying pairs of rows where only one condition and the decision outcome change values between the two rows.
> 
> Function Call and MC/DC are provided per line.
> 

When you are reading MCDC counters, do you want to show the branch counters as well in the source code view? (I've seen that you added new line mappings to the ´FileNode´. I tried to keep the memory footprint as small as possible in that class and combined line, branch and mutation coverages into a single value. This is now somewhat pointless with all the new values)

Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:122]: I will check the functioncall-coverage has good test data.

Please suggest another way to get the new way to get the coverages across the plugin instead of the expanded FileNode, and I can reimplement.

File: src/main/java/edu/hm/hafner/coverage/Metric.java [hunk 2]
```diff
@@ -32,6 +32,11 @@ public enum Metric {
     BRANCH(new ValuesAggregator()),
     INSTRUCTION(new ValuesAggregator()),
 
+    /* VectorCAST Coverage */
+    MCDC_PAIR(new ValuesAggregator()),
+    FUNCTION(new ValuesAggregator()),
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/coverage/Metric.java:37]: I think it would make sense to remove this metric as and reuse `METHOD`.

Comment (uhafner) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:37]: I think that I also found the problem why the `METHOD` metric does not work: It seems that your coverage.xml files do not map the lines to methods, the line nodes are just reported per file, and the methods are empty. I'll check if I can change the evaluation of the METHOD value on my side.

Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:37]: I can either change the metrics currently named "function-coverage" to something else "method-coverage" or I can add the lines to the method instead or along with the file.  Your choice.

Would you like me to rename the VectorCASTParser to ExtendedCoberturaParser?

File: src/main/java/edu/hm/hafner/coverage/Metric.java [hunk 3]
```diff
@@ -32,6 +32,11 @@ public enum Metric {
     BRANCH(new ValuesAggregator()),
     INSTRUCTION(new ValuesAggregator()),
 
+    /* VectorCAST Coverage */
+    MCDC_PAIR(new ValuesAggregator()),
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:34]: MCDC Pair describes using equivalence pair analysis.  Its a more accurate naming, but I can change it if you'd like.

[the remark this answers is not in the document]
Comment (uhafner) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:34]: No that is fine, in my lectures I am just using MCDC so I wondered about the suffix.

File: src/main/java/edu/hm/hafner/coverage/Metric.java (hunk 2 above)
Comment (uhafner) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:35]: > I can either change the metrics currently named "function-coverage" to something else "method-coverage" or I can add the lines to the method instead or along with the file. Your choice.

I am not sure if you can map the lines to the methods. They are not in the report file so I don't know how you can achieve that? 

Extract from your test file:
```
<methods>
            <method name="Get_Record" signature="(int)data_type" line-rate="1.0" branch-rate="1.0" complexity="2" functioncall-coverage="100.00% (2/2)" mcdcpair-coverage="100.00% (1/1)" function-coverage="100% (1/1)"/>
            <method name="Update_Record" signature="(int,data_type)" line-rate="1.0" branch-rate="0.6" complexity="2" functioncall-coverage="100.00% (1/1)" mcdcpair-coverage="0.00% (0/1)" function-coverage="100% (1/1)"/>
            <method name="Remove_Record" signature="(int)" line-rate="1.0" branch-rate="1.0" complexity="1" functioncall-coverage="100.00% (1/1)" function-coverage="100% (1/1)"/>
          </methods>
          <lines>
            <line number="45" hits="1" branch="true" condition-coverage="100.0% (1/1)">
              <conditions>
                <condition number="0" type="jump" coverage="100.0%"/>
              </conditions>
            </line>
            <line number="48" hits="1" branch="false"/>
            <line number="51" hits="1" branch="false"/>
            <line number="54" hits="1" branch="false" functioncall-coverage="100.0% (1/1)"/>
```
There is no way to see which lines are in which method, or am I missing something?

In Java Cobertura reports you have:

```
<method name="method">
<line number="48" hits="1" branch="false"/>
            <line number="51" hits="1" branch="false"/>
            <line number="54" hits="1" branch="false" functioncall-coverage="100.0% (1/1)"/>
</ method >
```

Comment (uhafner) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:35]: > Would you like me to rename the VectorCASTParser to ExtendedCoberturaParser?

No, the parser is actually a VectorCAST specific parser, this is fine. I think just the metrics and the model should not contain a tool name. Maybe we can use the correct Java capitalization which uses CamelCase even for short forms: `VectorCast`.

**Review by uhafner — CHANGES_REQUESTED** [review]
Thanks for this pull request, everything is well written and tested! 

The only bigger thing is the additional FUNCTION metric. I would be very happy if we can achieve the same results with the existing METHOD metric.

File: src/main/java/edu/hm/hafner/coverage/FileNode.java [hunk 4]
```diff
@@ -35,14 +35,23 @@
...
     private final NavigableMap<Integer, Integer> missedPerLine = new TreeMap<>();
 
+    // metrics for MC/DC pairs per line
+    private final NavigableMap<Integer, Integer> mcdcPairCoveredPerLine = new TreeMap<>();
+    private final NavigableMap<Integer, Integer> mcdcPairMissedPerLine = new TreeMap<>();
+
+    // metrics for function calls per line
+    private final NavigableMap<Integer, Integer> functionCallCoveredPerLine = new TreeMap<>();
+    private final NavigableMap<Integer, Integer> functionCallMissedPerLine = new TreeMap<>();
+
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/FileNode.java]: done

File: src/main/java/edu/hm/hafner/coverage/CoverageMetricsValues.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    private int missed;
+    private int total;
+    
+    /**
+     * Creates a new {@link CoverageMetricsValues} instance from the values for covered/missing. 
+     * These values represent the coveraged and missed metrics
+     *
+     * @param covered  the coverage count
+     * @param missed   the missed count
+     *
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/CoverageMetricsValues.java]: done

File: src/main/java/edu/hm/hafner/coverage/CoverageMetricsValues.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+     */
+
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/CoverageMetricsValues.java]: done

File: src/main/java/edu/hm/hafner/coverage/CoverageMetricsValues.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
+package edu.hm.hafner.coverage;
+
+/**
+ * Used to simplify FileNode.
+ *
+ * @author Tim Schneider
+ */
+public class CoverageMetricsValues {
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/CoverageMetricsValues.java]: done

File: src/main/java/edu/hm/hafner/coverage/FileNode.java [hunk 8]
```diff
@@ -521,6 +587,46 @@ public FileNode addCounters(final int lineNumber, final int covered, final int m
...
+     *         the line number to add the counters for
+     * @param covered
+     *         the number of covered items
+     * @param missed
+     *         the number of missed items
+     *
+     * @return this instance
+     */
+    @CanIgnoreReturnValue
+    public FileNode addMCDCPairCounters(final int lineNumber, final int covered, final int missed) {
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/FileNode.java]: done

File: src/main/java/edu/hm/hafner/coverage/Metric.java (hunk 2 above)
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:35]: done

File: src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    private Coverage fromFunctionCallCoverage(final String functionCallCoverageAttribute) {
+        var matcher = BRANCH_PATTERN.matcher(functionCallCoverageAttribute);
+        if (matcher.matches()) {
+            return new CoverageBuilder().withMetric(Metric.FUNCTION_CALL)
+                    .withCovered(matcher.group("covered"))
+                    .withTotal(matcher.group("total"))
+                    .build();
+        }
+        return Coverage.nullObject(Metric.FUNCTION_CALL);
+    }
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java]: done

File: src/test/java/edu/hm/hafner/coverage/FileNodeTest.java [hunk 10]
```diff
@@ -74,11 +74,12 @@ void shouldGetFilePath() {
 
     @Test
     void shouldReadOldVersion() {
-        byte[] restored = readAllBytes("version-0.21.0.ser");
+        // byte[] restored = readAllBytes("version-0.21.0.ser");
 
         var serializable = (FileNode)createSerializable();
         serializable.setRelativePath(TreeString.valueOf(StringUtils.EMPTY));
-        assertThatRestoredInstanceEqualsOriginalInstance(serializable, restore(restored));
+        // version-0.21.0.ser needs to be updated
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/test/java/edu/hm/hafner/coverage/FileNodeTest.java]: done

File: src/test/java/edu/hm/hafner/coverage/parser/VectorCASTParserTest.java [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+        long missedLines = 0;
+        long coveredLines = 0;
+        for (Node node : nodes) {
+            var lineCoverage = (Coverage) node.getValue(LINE).get();
+            missedLines = missedLines + lineCoverage.getMissed();
+            coveredLines = coveredLines + lineCoverage.getCovered();
+        }
+
+        assertThat(missedLines).isEqualTo(28);
+        assertThat(coveredLines).isEqualTo(52);
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/test/java/edu/hm/hafner/coverage/parser/VectorCASTParserTest.java]: done

File: src/test/java/edu/hm/hafner/coverage/parser/VectorCASTParserTest.java [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+        assertThat(missedBranches).isEqualTo(92);
+        
+        assertThat(coveredMcdcPairs).isEqualTo(24);
+        assertThat(missedMcdcPairs).isEqualTo(35);
+        
+        assertThat(coveredFunctions).isEqualTo(21);
+        assertThat(missedFunctions).isEqualTo(9);
+        
+        assertThat(coveredFunctionCalls).isEqualTo(62);
+        assertThat(missedFunctionCalls).isEqualTo(17);
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/test/java/edu/hm/hafner/coverage/parser/VectorCASTParserTest.java]: done

File: src/main/java/edu/hm/hafner/coverage/FileNode.java [hunk 13]
```diff
@@ -521,6 +587,46 @@ public FileNode addCounters(final int lineNumber, final int covered, final int m
         return this;
     }
 
+    /**
+     * Add the coverage counters for the specified line.
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/FileNode.java]: done

File: src/main/java/edu/hm/hafner/coverage/Metric.java [hunk 14]
```diff
@@ -32,6 +32,11 @@ public enum Metric {
     BRANCH(new ValuesAggregator()),
     INSTRUCTION(new ValuesAggregator()),
 
+    /* VectorCAST Coverage */
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/Metric.java]: done

File: src/main/java/edu/hm/hafner/coverage/Metric.java (hunk 1 above)
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:120]: done

File: src/main/java/edu/hm/hafner/coverage/Node.java [hunk 15]
```diff
@@ -872,5 +872,7 @@ private Optional<Node> filterTreeByMapping(final Function<Node, Optional<Node>>
      *
      * @return {@code true} if this node is an aggregation node, {@code false} otherwise
      */
-    public abstract boolean isAggregation();
+    public boolean isAggregation() {
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/Node.java]: i was trying to remove public content from FileNode.  I reversed this change

File: src/main/java/edu/hm/hafner/coverage/parser/CoberturaParser.java [hunk 16]
```diff
@@ -95,7 +95,7 @@ protected ModuleNode parseReport(final Reader reader, final String fileName, fin
         }
     }
 
-    private boolean readModule(final XMLEventReader eventReader, final ModuleNode root,
+    protected boolean readModule(final XMLEventReader eventReader, final ModuleNode root,
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/parser/CoberturaParser.java:98]: yes, not as many as I thought.  Cleaned up.

File: src/main/java/edu/hm/hafner/coverage/FileNode.java [hunk 17]
```diff
@@ -193,16 +243,44 @@ private void mergeCounters(final FileNode otherFile) {
...
+            addValue(mcdcPairValue);
+        }
+
+        var functionCallValue = functionCallCoverage.build();
+        if (functionCallValue.isSet()) {
+            addValue(functionCallValue);
+        }
+    }
+            
+    void mergeLeftRight(final int line, final int leftCovered, final int leftMissed, final int rightCovered, final int rightMissed,
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/coverage/FileNode.java:257]: When not used somewhere else:
```suggestion
    private void mergeLeftRight(final int line, final int leftCovered, final int leftMissed, final int rightCovered, final int rightMissed, 
```

File: src/main/java/edu/hm/hafner/coverage/Metric.java (hunk 2 above)
Comment (uhafner) [reply src/main/java/edu/hm/hafner/coverage/Metric.java:35]: Do you need some help with the model so we can eliminate the duplicate enum `FUNCTION`? It would be very helpful if we can omit this value and use the `METHOD` coverage.

File: src/main/java/edu/hm/hafner/coverage/parser/VectorCastParser.java [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+import edu.hm.hafner.coverage.Node;
+import edu.hm.hafner.util.FilteredLog;
+import java.util.Map;
+import java.util.HashMap;
+
+/**
+ * Parses VectorCAST reports into a hierarchical Java Object Model.
+ *
+ * @author Melissa Bauer
+ * @author Ullrich Hafner
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/coverage/parser/VectorCastParser.java:24]: Your name?

File: src/main/java/edu/hm/hafner/coverage/parser/VectorCastParser.java [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+     */
+    public VectorCastParser(final ProcessingMode processingMode) {
+        super(processingMode);
+    }
+        
+    private Coverage processClassMethodStart(final StartElement nextElement, final Coverage functionCoverage) {
+        var localFunctionCoverage = functionCoverage;
+        
+        if (nextElement.getName().equals(METHOD)) {
+        // METHOD.equals(nextElement.getName())) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/coverage/parser/VectorCastParser.java:66]: Remove?

**Review by uhafner — COMMENTED** [review]
Thanks for fixing all those tasks! The only thing I would like to change is the duplicate enum FUNCTION vs. METHOD. Do you think that we can eliminate it? I think I almost got rid of it a couple of weeks ago but then dropped my changes... If you need help, I can push a draft of these changes again...

File: src/main/java/edu/hm/hafner/coverage/FileNode.java (hunk 17 above)
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/FileNode.java]: Complete

File: src/main/java/edu/hm/hafner/coverage/parser/VectorCastParser.java (hunk 19 above)
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/parser/VectorCastParser.java]: Complete

File: src/main/java/edu/hm/hafner/coverage/FileNode.java [hunk 20]
```diff
@@ -35,14 +35,23 @@
  *
  * @author Ullrich Hafner
  */
-@SuppressWarnings({"PMD.GodClass", "PMD.CyclomaticComplexity", "PMD.CouplingBetweenObjects"})
+// Added ExcessivePublicCount - This class has a bunch of public methods and attributes
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/FileNode.java]: complete

File: src/main/java/edu/hm/hafner/coverage/parser/VectorCastParser.java (hunk 18 above)
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/parser/VectorCastParser.java:24]: complete

**Comment by TimSVector** [comment]
> Thanks for fixing all those tasks! The only thing I would like to change is the duplicate enum FUNCTION vs. METHOD. Do you think that we can eliminate it? I think I almost got rid of it a couple of weeks ago but then dropped my changes... If you need help, I can push a draft of these changes again...

Please push any suggested changes for this.  I cannot figure it out.  I think what's tripping me up is Method depends on Children and Function is a Value (2/4 50%)

```     
    FUNCTION(new ValuesAggregator()),
    METHOD(new LocOfChildrenEvaluator()),
```

**Comment by uhafner** [comment]
Yes, this exactly was the problem. I fixed that code now. When a local value is present, then we do not need to derive the "container" metrics from the line coverage.

I pushed my changes here: https://github.com/jenkinsci/coverage-model/commit/d9a047cdb41c84c8a6c69b05ca9091b91e048363

If you prefer, I can push the changes back to your repo?

**Comment by TimSVector** [comment]
> Thanks for this pull request, everything is well written and tested!
> 
> The only bigger thing is the additional FUNCTION metric. I would be very happy if we can achieve the same results with the existing METHOD metric.

Made the changes and it looks like coverage model is working as expected.  Changes being made to coverage-plugin to go along with this changes

File: src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+ * Parses VectorCAST reports into a hierarchical Java Object Model.
+ *
+ * @author Melissa Bauer
+ * @author Ullrich Hafner
+ */
+@SuppressWarnings({"checkstyle:ClassDataAbstractionCoupling", "PMD.GodClass"})
+public class VectorCASTParser extends CoberturaParser {
+    private static final long serialVersionUID = 598117573006409816L;
+ 
+    private static final Pattern BRANCH_PATTERN = Pattern.compile(".*\\((?<covered>\\d+)/(?<total>\\d+)\\)");
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java]: complete

File: src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+                if (CLASS.equals(endElement.getName()) || METHOD.equals(endElement.getName())) {
+                    processClassMethodEnd(node, coverageMap);
+                    return;
+                }
+            }
+        }
+        throw createEofException(fileName);
+    }
+    
+    private boolean isVectorCASTCoverage(final StartElement line, final QName attribute) {
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java]: Complete - OBE

File: src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+        boolean retVal = false;
+        
+        Attribute cov = line.getAttributeByName(attribute);
+        
+        if (cov != null) {
+            retVal = true;
+        } 
+        return retVal;
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java]: Complete - OBE

File: src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+ * @author Melissa Bauer
+ * @author Ullrich Hafner
+ */
+@SuppressWarnings({"checkstyle:ClassDataAbstractionCoupling", "PMD.GodClass"})
+public class VectorCASTParser extends CoberturaParser {
+    private static final long serialVersionUID = 598117573006409816L;
+ 
+    private static final Pattern BRANCH_PATTERN = Pattern.compile(".*\\((?<covered>\\d+)/(?<total>\\d+)\\)");
+
+    private static final Coverage DEFAULT_BRANCH_COVERAGE       = new CoverageBuilder(Metric.BRANCH).withCovered(2).withMissed(0).build();
```
[the remark this answers is not in the document]
Comment (TimSVector) [reply src/main/java/edu/hm/hafner/coverage/parser/VectorCASTParser.java]: I removed it an its using the CoberturaParser.java call for readBranchCoverage().  Cobertura also has :

`private static final Coverage DEFAULT_BRANCH_COVERAGE = `
`        new CoverageBuilder(Metric.BRANCH).withCovered(2).withMissed(0).build();`

So you may want to look at that instead.

**Review by TimSVector — COMMENTED** [review]
All requested changes made.  Ready to move forward

**Comment by TimSVector** [comment]
> Thanks for this pull request, everything is well written and tested!
> 
> The only bigger thing is the additional FUNCTION metric. I would be very happy if we can achieve the same results with the existing METHOD metric.

Method metrics works fine now and function coverage has been replaced

**Review by uhafner — APPROVED** [review]

**Comment by uhafner** [comment]
Superseded by #116

**Comment by uhafner** [comment]
Thanks for this PR, I will test it with the coverage plugin now and create a new release!
